### PR TITLE
#9: add automated testing of basic cases

### LINKED ISSUE #9: Add automated testing of basic cases
To start, use the same cases as the one that are tested in LBAF, but now using the MILP configuration. Extract the optimum and verify we obtain the same results.

### PR DESCRIPTION
**Pull request by maxime-bfsquall**
resolve #9 

Reviews the code to separete each class for more modularity
Create of two scripts 'ccm-milp_problem.py' and 'ccm_milp_solver.py' to seperate functionnalities
Review code to be pylint compliant
Add "Code Quality" into the CI: 
- Thanks to Tox, a script is launched to test all the examples with 3 differents solvers (PULP_CBC_CMD, COIN_CMD and GLPK_CMD)
- Add pylint test
- Report test results

### CODE REVIEW

File: examples/illustration_1.py
Comment (ppebay) [inline examples/illustration_1.py:1]: What we really want at this stage is to use the existing `SyntheticBlock` cases -- no other examples needed at this point -- , and to test it with a variety of configurations, in particular:
- load-only (COMCP) alon
- load-only + cluster-preserving
- load-only + memory bound
- FWMP with alpha = 1, all other coeffs at 0, which should amount to exactly the same results as above
- FWMP with alpha =1 and some non zero values of beta, gamma, delta

Ideally this should be synchronized with what has already been done to test LBAF, for the same case, with @cwschilly

**Review by ppebay — CHANGES_REQUESTED** [review]

File: src/CcmMilp/Configuration.py [hunk 1]
```diff
@@ -42,66 +42,109 @@ def __init__(
...
+                        'file': '1-load-only.yaml',
+                        'regexp': [
+                            'Optimal - objective value 2.00000000',
+                            'Objective:  OBJ = 2 (MINimum)'
+                        ]
+                    },
+                    {
+                        'file': '1-load-only-and-cluster.yaml',
+                        'regexp': [
+                            'Optimal - objective value 2.00000000',
```
Comment (ppebay) [inline src/CcmMilp/Configuration.py:105]: Are you sure that the optimal value found in this case isn't 2.5 (instead of 2.0)?

File: src/CcmMilp/Configuration.py [hunk 2]
```diff
@@ -42,66 +42,109 @@ def __init__(
...
+                    },
+                    {
+                        'file': '1-fwmp-with-alpha.yaml',
+                        'regexp': [
+                            'Optimal - objective value 2.00000000',
+                            'Objective:  OBJ = 2 (MINimum)'
+                        ]
+                    },
+                    {
+                        'file': '1-fwmp-with-alpha-beta.yaml',
```
Comment (ppebay) [inline src/CcmMilp/Configuration.py:124]: What is this case testing exactly with `beta`?

File: src/CcmMilp/Configuration.py (hunk 1 above)
Comment (maxime-bfsquall) [reply src/CcmMilp/Configuration.py:105]: Maybe my understanding of the configuration is wrong.

Yes, we found **2.0** with the current configuration : 
`is_fwmp: False
alpha: 0
beta: 0
gamma: 0
delta: 0
bounded_memory: False
preserve_clusters: True`

We found **0.0** when we set is_fwmp to True:
`is_fwmp: True
alpha: 0
beta: 0
gamma: 0
delta: 0
bounded_memory: False
preserve_clusters: True`

We found **2.5** when we set is_fwmp to True and alpha to 1: 
`is_fwmp: True
alpha: 1
beta: 0
gamma: 0
delta: 0
bounded_memory: False
preserve_clusters: True`

File: src/CcmMilp/Configuration.py (hunk 2 above)
Comment (maxime-bfsquall) [reply src/CcmMilp/Configuration.py:124]: I just reviewed the configuration, it tests: 
`is_fwmp: True
alpha: 1
beta: 1
gamma: 0
delta: 0
bounded_memory: False
preserve_clusters: False`

And the objective value found is 4.0

File: src/CcmMilp/Configuration.py (hunk 1 above)
Comment (ppebay) [reply src/CcmMilp/Configuration.py:105]: A bug was introduced, because it is **not** possible to find 2.0 when clusters are preserved.  It seems that the value of `True` for `preserve_clusters` is not used with this version of the code. We need to look into this.

When `preserve_clusters` is `True` we cannot achieve anything better than 2.5 with the COMCP (FWMP=`False`) model.

Comment (ppebay) [reply src/CcmMilp/Configuration.py:105]: @maxime-bfsquall I know what happened! I introduced an incorrect indentation in the previous branch that got merged (and because I did not manually test the COMCP I did not notice it!):
Screenshot 2024-09-10 at 07 32 36

BTW this is a major flaw a Python, this dependency on indentation (which I had  criticized from the first day when I discovered that language almost 2 decades ago!)

I am pushing the fix as part of this branch.

So, please note:
- you should now find 2.5 for this case
- when running in COMCP mode (i.e. when FWMP is set to `False`) you do **not** need to set the values of the `alpha`, `beta`, `gamma`, and `delta` coefficients, as those are automatically set to 1, 0, 0, and 0 automatically.

Comment (ppebay) [reply src/CcmMilp/Configuration.py:107]: @maxime-bfsquall I pushed the correction and I confirm that the results are now correct in `COMCP` mode as well:

Screenshot 2024-09-10 at 07 44 41

So for this case the parameters should be as follows:
```
is_fwmp: False bounded_memory: False preserve_clusters: True
```
and the computed optimum should be **2.5**.

It will therefore match the correct result of **2.5** that you are already finding for the full model with these parameters:
```
is_fwmp: True alpha: 1 beta: 0 gamma: 0 delta: 0 bounded_memory: False preserve_clusters: True
```
as it is exactly the same, as the COMCP example above.

Comment (maxime-bfsquall) [reply src/CcmMilp/Configuration.py]: I changed the objective value to the expected value **2.5** for "load-only + cluster-preserving"
I also removed the coefficient values ​​in the test configuration when their value matches the default

Comment (ppebay) [reply src/CcmMilp/Configuration.py]: Alright, so here is now the other case that I wanted to discuss:
```
is_fwmp: True alpha: 0 beta: 0 gamma: 0 delta: 0 bounded_memory: False preserve_clusters: True
```
for this case you mentioned that you were finding 0.0, which is technically correct -- but it's not surprising indeed, as it amounts to a combination that is always equal to 0. Therefore, the optimum can only be 0!

At first glance I wanted to ask you to remove this case, but after having thought a bit more about it, I think it's best to keep it, but just to give it a different name, e.g., `null-case`. It's better to test this corner case as well.

That being said, we should then **also** test the following case
```
is_fwmp: True alpha: 1.0 beta: 0.0 gamma: 0.0 delta: 0.0 bounded_memory: False preserve_clusters: False
```
and make sure that find **2.0**. This is to verify that this FWMP, which is equivalent to the simplest COMCP case, also reports a correct value.

Comment (maxime-bfsquall) [reply src/CcmMilp/Configuration.py]: I added `null-case` into the test process.

The case :  
`is_fwmp: True alpha: 1.0 beta: 0.0 gamma: 0.0 delta: 0.0 bounded_memory: False preserve_clusters: False` 
is already tested under the config 'fwmp-with-alpha'

Comment (ppebay) [reply src/CcmMilp/Configuration.py]: Ok great, resolving this conversation. Thanks @maxime-bfsquall

File: src/CcmMilp/Configuration.py (hunk 2 above)
Comment (ppebay) [reply src/CcmMilp/Configuration.py:126]: Alright @maxime-bfsquall before testing the example your mentioned above (which combines the **compute** (`alpha`) and **off-node communication** (`beta`) terms, let's first make sure that we correctly resolve the latter only (`beta`). Can you thus please add the following `off-node-communication-only` case:
```
is_fwmp: True alpha: 0.0 beta: 1.0 gamma: 0.0 delta: 0.0 bounded_memory: False preserve_clusters: False
```
and make sure that we find the following optimum: **0.0**?

Thanks

File: src/CcmMilp/Configuration.py
Comment (ppebay) [inline src/CcmMilp/Configuration.py:1]: Please rename this directory name **not** using mixed/camel case as it's not consistent with the fact that these are all initials.
I think we could use either `CCM_MILP` or `ccm_milp`

File: src/CcmMilp/Generator.py
Comment (ppebay) [inline src/CcmMilp/Generator.py:1]: Please do not capitalize Python file names; use `generator.py` instead. Same remark for all other Python files here.

**Review by ppebay — CHANGES_REQUESTED** [review]

File: README.md
Comment (ppebay) [inline README.md:1]: Can you please an `## Expected Results:` section, providing a bulleted-list of all the cases that are currently in CI, with:
* their configurations (i.e. `FWMP=False, preserve_clusters=False` etc.
* their expected correct optimization results

This is for ease of comparison with temperedLB & LBAF.

File: examples/data/illustration_1.py
Comment (ppebay) [inline examples/data/illustration_1.py:1]: Why is this called called `illustration_1`? Does it correspond to a know case (or published elsewhere)?

File: src/ccm_milp/configuration.py
Comment (ppebay) [inline src/ccm_milp/configuration.py:1]: Would there be a different way to implement this, so GitHub does not complain with the "too few public methods" nonsensical style warning?

File: src/ccm_milp/generator.py [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+        else: # if self.problem.status == pulp.LpStatusOptimal
+            print("# No solution found")
+
+    def setup_milp(self):
+        """ Solve a minimization of a mixed-integer linear program. """
+        # instantiante LP minimization problem
+        self.problem = pulp.LpProblem("CCM_MILP", pulp.LpMinimize)
+
+        # Number of ranks
+        self_i = self.i
```
Comment (ppebay) [inline src/ccm_milp/generator.py:171]: What is the point of using these `self_` local variables? Why not use the `self.` member variables?

File: tests/config/1-load-only.yaml
Comment (ppebay) [inline tests/config/1-load-only.yaml:1]: Please make sure to add also the COMCP case (`is_fwmp: False`) with `preserve_clusters` set to `True` also. Expected value is 2.5

**Review by ppebay — CHANGES_REQUESTED** [review]

File: src/ccm_milp/generator.py (hunk 3 above)
Comment (maxime-bfsquall) [reply src/ccm_milp/generator.py:171]: I removed them and use `self` instead.
I had changed the variable names to be pylint compliant and had not remove them because I had not checked if these variables were modified in the algorithm

File: tests/config/1-load-only.yaml
Comment (maxime-bfsquall) [reply tests/config/1-load-only.yaml:1]: Yes, it is (1-load-only-and-cluster.yml)

File: examples/data/illustration_1.py
Comment (maxime-bfsquall) [reply examples/data/illustration_1.py:1]: This example was already here when I started. Maybe @tlamonthezie can help to understand ?

File: src/ccm_milp/configuration.py
Comment (maxime-bfsquall) [reply src/ccm_milp/configuration.py:1]: We can disable these messages globally or on a specific code using comments

File: README.md [hunk 4]
```diff
@@ -1,15 +1,85 @@
...
+
+* Load only and-cluster
+    * Configuration: `is_fwmp = False, preserve_clusters = True`
+    * Optimal objective value `2.50000000`
+
+* Load only and-memory-bound
+    * Configuration: `is_fwmp = False, bounded_memory: True`
+    * Optimal objective value `2.00000000`
+
+* Fwmp with alpha
```
Comment (ppebay) [inline README.md:71]: Please capitalize (FWMP)

File: examples/data/illustration_1.py
Comment (ppebay) [reply examples/data/illustration_1.py:1]: No big deal, we can discuss at meeting, resolving. Thanks

File: src/ccm_milp/configuration.py
Comment (ppebay) [reply src/ccm_milp/configuration.py:1]: Yes but I would rather not disable them, if there is a way to genuinely improve the code (as in: make it more Pythonic)

**Review by ppebay — CHANGES_REQUESTED** [review]
Only minuscule changes requested now, this is almost complete

File: README.md
Comment (maxime-bfsquall) [reply README.md:1]: This section was added at the end of the README

**Review by ppebay — APPROVED** [review]
We have converged! Thanks @maxime-bfsquall
